### PR TITLE
feat(components): create anchor links

### LINKED ISSUE #904: Need ability to create anchor links
## Description
As a content editor I want to create anchor links on sections/copy/etc. so that I can use these urls elsewhere to drop users directly to this content.

The use case here is the anchors were added to the h2's faq sections.

https://www.microlax.com.mx/faq#informaci%C3%B3n-general-acerca-de-microlax-sup-sup

 

 
## Acceptance Criteria
AC1. There is an HOC that adds the following effect to a component
- Adds a local context button, in the context group anchor and merges down
- This button opens a form that allows for the entering of on id which has a placeholder value of "Descriptive ID"
- This button label is "add" if there is no id, and "edit" if there is an id
- The button icon is "Tag" icon.
- This id is then applied to the component if specified 
- For form validation/normalization of id:
-- If a # is entered, the # is removed automatically before saving.
-- The id is transformed (spaces to dashes otherwise encode it) to create valid anchor , the transformed anchor id value appears underneath the ID input field of the form. 

AC2. The test site has the standard H2 wrapped with the above HOC (scoped to RTE H2 and Accordion TItle H2)

AC3. If the anchor ID is empty then nothing is added

AC4. If this HOC is applied to a component that already has data, the existing data is still rendered correctly (the anchor id value is added as sidecar)

AC5. Address Bug "Seach is overriding the #anchor link functionality" is resolved

### PR DESCRIPTION
**Pull request by awm086**
## Changes
- New `asBodilessAnchor` HOC that allows site editor to add an `id` to the wrapped component so that it may work as an anchor link

## Test Instructions
- Test adding anchor links to the Headers on the `/anchor` page and navigating to it.

## Related Issues
Closes #904

### CODE REVIEW

File: examples/test-site/src/data/pages/anchor/index.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import React from 'react';
+import { flow } from 'lodash';
+import { graphql } from 'gatsby';
+import { Page } from '@bodiless/gatsby-theme-bodiless';
+import { asBodilessAnchor, asBodilessLink } from '@bodiless/components';
+import { A, asToken, H1, H2, Span } from '@bodiless/fclasses';
```
Comment (VancheeZze) [inline examples/test-site/src/data/pages/anchor/index.tsx:19]: `asToken` is not used anywhere, it generates the lint issue

File: packages/bodiless-components/src/Anchor/asBodilessAnchor.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          placeholder="Descriptive ID"
+        />
+        {errors && errors.id && (
+          <ComponentFormWarning>{errors.id}</ComponentFormWarning>
+        )}
+      </>
+    );
+  },
+};
+const asBodilessAnchor: AsBodiless<Props, Data> = (
```
Comment (VancheeZze) [inline packages/bodiless-components/src/Anchor/asBodilessAnchor.tsx:69]: I'd add a TSDocs to describe `asBodilessAnchor`

File: packages/bodiless-components/src/Anchor/asBodilessAnchor.tsx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+const anchorOptions: BodilessOptions<Props, Data> = {
+  icon: 'local_offer',
+  groupLabel: 'Anchor',
+  label: 'Add',
+  name: 'anchor-settings',
+  global: false,
+  local: true,
+  renderForm: ({ formState, scope }) => {
+    // eslint-disable-next-line no-useless-escape
+    const isValidHtmlId = (id : string) => (/^[A-Za-z]+[\w\-\:\.]*$/.test(id));
```
Comment (VancheeZze) [inline packages/bodiless-components/src/Anchor/asBodilessAnchor.tsx:43]: Do we want `isValidHtmlId` defined inside `renderForm`? Maybe we can move it outside as a standalone method outside of `anchorOptions`?

**Review by VancheeZze — APPROVED** [review]

**Comment by irkhodzka** [comment]
Dear @VancheeZze ,
please find 
**issue 1:**
"#" sign entered to "Descriptive ID" is not removed automatically.
**STR:**
1. Go to the /anchor page. 
2. Switch to edit mode. 
3. Click on any component.
4. Click on "Add"/"Edit" button.
5. Enter #value to the "Descriptive ID" field (e.g. - #overview)

**Actual Result:**
Validation message appears saying that value is invalid.

**Expected Result:** 
AC1 _-- If a # is entered, the # is removed automatically before saving._

**Comment by irkhodzka** [comment]
**issue2:**
Spaces in "Descriptive id" value are not transformed to dashes underneath the ID input field of the form.
**STR:**
1. Go to the /anchor page. 
2. Switch to edit mode. 
3. Click on any component.
4. Click on "Add"/"Edit" button.
5. Enter any value with spaces to the "Descriptive ID" field (e.g. - product overview)

**Actual Result:**
Validation message appears saying that value is invalid.

**Expected Result:** 
AC1: _Spaces in "Descriptive id" value are transformed to dashes and appear underneath the ID input field of the form._

**Comment by VancheeZze** [comment]
@irkhodzka 
**issue1** and **issue2** have been fixed. The demo page is updated.

Thank you.

**Review by beliayeu — APPROVED** [review]

**Comment by irkhodzka** [comment]
Dear @VancheeZze ,
please see below the **issue 3** found:
After removing already saved descriptive ID the empty value can not be saved.

**STR**:
1. Go to the /anchor page. 
2. Click on any component with anchor link being set to it.
3. Delete anchor link
4. Click "V" to save empty value.

**Actual Result**: "V" button does not respond on clicking. User is navegated to the page containing code called "TypeError: Cannot read property 'replace' of undefined.
https://user-images.githubusercontent.com/81805756/114995440-7b942c00-9ea6-11eb-9145-56953d62c17a.mp4

**Expected Result:** After deleting value from Descriptive ID an empty field is saved, Anchor link is absent on the element.

**Comment by VancheeZze** [comment]
@irkhodzka 

**issue 3** has been fixed. Envs are redeployed.

Thank you!

**Review by beliayeu — APPROVED** [review]

**Review by awm086 — COMMENTED** [review]
approved 👍

**Review by VancheeZze — APPROVED** [review]

**Pull request merged**